File: water-loggers/WaterTracker.kt
import java.util.Scanner

const val DAILY_GOAL = 2000.0 // Daily water intake goal in milliliters
const val WELCOME_MESSAGE = "Welcome to the Water Tracker App!"
const val EXIT_PROMPT = "Enter the amount of water you drank in ml (or type 'exit' to quit): "
const val INVALID_INPUT_MESSAGE = "Invalid input. Please enter a number."
const val POSITIVE_NUMBER_MESSAGE = "Please enter a positive number."
const val GOAL_REACHED_MESSAGE = "Congratulations! You've reached your daily goal!"
const val KEEP_GOING_MESSAGE = "Keep going! You still need %s ml to reach your goal."
const val THANK_YOU_MESSAGE = "Thank you for using the Water Tracker App! Your total intake today was %s ml."

fun main() {
    val scanner = Scanner(System.`in`)
    var totalWaterIntake = 0.0

    println(WELCOME_MESSAGE)
    println("Your daily goal is $DAILY_GOAL ml.")

    while (true) {
        println(EXIT_PROMPT)
        val input = scanner.nextLine()

        if (input.lowercase() == "exit") {
            break
        }

        val waterIntake = validateInput(input)
        if (waterIntake != null) {
            totalWaterIntake += waterIntake
            println("Total water intake: $totalWaterIntake ml")
            checkGoal(totalWaterIntake, DAILY_GOAL)
        }
    }

    println(THANK_YOU_MESSAGE.format(totalWaterIntake))
}

fun validateInput(input: String): Double? {
    return try {
        val waterIntake = input.toDouble()
        if (waterIntake < 0) {
            println(POSITIVE_NUMBER_MESSAGE)
            null
        } else {
            waterIntake
        }
    } catch (e: NumberFormatException) {
        println(INVALID_INPUT_MESSAGE)
        null
    }
}

fun checkGoal(totalIntake: Double, goal: Double) {
    if (totalIntake < goal) {
        println(KEEP_GOING_MESSAGE.format(goal - totalIntake))
    } else {
        println(GOAL_REACHED_MESSAGE)
    }
}

File: water-loggers/water_tracker.rs
use std::io;

const WELCOME_MESSAGE: &str = "Welcome to the Water Tracker App!";
const PROMPT_MESSAGE: &str = "Enter the amount of water you drank in liters (type 'exit' to quit):";
const POSITIVE_NUMBER_MESSAGE: &str = "Please enter a positive number.";
const INVALID_INPUT_MESSAGE: &str = "Invalid input. Please enter a number or 'exit' to quit.";
const TOTAL_INTAKE_MESSAGE: &str = "Your total water intake for today is: {:.2} liters.";

fn main() {
    let mut total_water: f32 = 0.0;

    println!("{}", WELCOME_MESSAGE);
    println!("{}", PROMPT_MESSAGE);

    loop {
        let input = get_user_input();

        if input.eq_ignore_ascii_case("exit") {
            break;
        }

        match parse_input(&input) {
            Ok(amount) if amount > 0.0 => {
                total_water += amount;
                println!("Added {:.2} liters. Total water intake: {:.2} liters.", amount, total_water);
            }
            Ok(_) => {
                println!("{}", POSITIVE_NUMBER_MESSAGE);
            }
            Err(_) => {
                println!("{}", INVALID_INPUT_MESSAGE);
            }
        }
    }

    println!(TOTAL_INTAKE_MESSAGE, total_water);
}

fn get_user_input() -> String {
    let mut input = String::new();
    io::stdin().read_line(&mut input).expect("Failed to read line");
    input.trim().to_string()
}

fn parse_input(input: &str) -> Result<f32, std::num::ParseFloatError> {
    input.parse::<f32>()
}
